bridgerelayer/cmd/chains/ethereum/router.go

```diff
@@ -2,50 +2,75 @@ package ethereum
 
 import (
 	"bytes"
-	"context"
-	"log"
 
-	ethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/common"
+	ctypes "github.com/ethereum/go-ethereum/core/types"
+	log "github.com/sirupsen/logrus"
 
+	keybase "github.com/snowfork/polkadot-ethereum/bridgerelayer/cmd/keybase/ethereum"
 	"github.com/snowfork/polkadot-ethereum/bridgerelayer/cmd/types"
+	"github.com/snowfork/polkadot-ethereum/prover"
 )
 
-// EthRouter ...
-type EthRouter struct{}
+// Router packages raw event data as Packets and relays them to the bridge
+type Router struct {
+	keybase *keybase.Keypair
+}
+
+// NewRouter initializes a new instance of Router
+func NewRouter(keybase *keybase.Keypair) Router {
+	return Router{
+		keybase: keybase,
+	}
+}
 
-// BuildPacket ...
-func (er *EthRouter) BuildPacket(tx ethTypes.Transaction, block ethTypes.Block) (types.Packet, error) {
-	chainID, err := client.NetworkID(context.Background())
+// Route packages tx data as a packet and relays it to the bridge
+func (er Router) Route(eventData types.EventData) error {
+	packet, err := er.buildPacket(eventData.Contract, eventData.Data)
 	if err != nil {
-		log.Fatal(err)
-		return types.Packet{}, err
+		return err
 	}
 
-	receipt, err := client.TransactionReceipt(context.Background(), tx.Hash())
+	err = er.sendPacket(packet)
 	if err != nil {
-		log.Fatal(err)
-		return types.Packet{}, err
+		return err
 	}
 
-	var receiptBuf bytes.Buffer
-	receipt.EncodeRLP(&receiptBuf)
+	return nil
+}
+
+// BuildPacket builds a data packet from tx data
+func (er Router) buildPacket(id common.Address, eLog ctypes.Log) (types.Packet, error) {
+	appAddress := id.Bytes()
+	var appID [32]byte
+	copy(appID[:], appAddress)
 
-	// Transaction data
-	txData := types.NewEthTxData(chainID.Bytes(), block.Hash().Bytes(),
-		tx.Hash().Bytes(), tx.Data(), receiptBuf.Bytes())
+	// RLP encode event log's Address, Topics, and Data
+	var buff bytes.Buffer
+	err := eLog.EncodeRLP(&buff)
+	if err != nil {
+		return types.Packet{}, err
+	}
 
-	// Message
-	message := types.NewMessage(txData, []byte{})
+	// Generate a proof by signing a hash of the encoded data
+	proof, err := prover.GenerateProof(buff.Bytes(), er.keybase.PrivateKey())
```

Comment:
A general comment on private key handling: It would be more secure to just accept a `Signer` interface as opposed to a private key and likewise wrap the private key in a struct that only exposes the signer interface. That way, the consumer could never access any secret data. That prevents accidental exposure and prepares our codebase for HSMs/ledger devices and remote signing.